File: rust/2015/day18b/src/main.rs
use std::env;
use std::fs::File;
use std::io::{self, BufRead};
use std::process;

type Grid = Vec<Vec<u32>>;

const NUM_STEPS: u32 = 100;

fn usage(progname: &str) {
    eprintln!("usage: {progname} <input file>");
    process::exit(1);
}

fn parse_file(filename: &str) -> io::Result<Grid> {
    let file = File::open(filename)?;
    let reader = io::BufReader::new(file);
    let mut grid: Grid = vec![];
    for line_result in reader.lines() {
        let line = line_result?;
        let row: Vec<u32> = line
            .chars()
            .map(|x| match x {
                '#' => 1,
                '.' => 0,
                _ => panic!("malformed input: characters must be # or ."),
            })
            .collect();
        grid.push(row);
    }
    Ok(grid)
}

fn count_neighbors(grid: &Grid, row: usize, col: usize) -> u32 {
    let deltas: [(isize, isize); 8] = [
        (-1, -1),
        (-1, 0),
        (-1, 1),
        (0, -1),
        (0, 1),
        (1, -1),
        (1, 0),
        (1, 1),
    ];
    let num_rows = grid.len();
    let num_cols = grid[0].len();
    deltas
        .iter()
        .map(|(i, j)| (row as isize + i, col as isize + j))
        .filter(|(r, c)| *r >= 0 && *r < num_rows as isize && *c >= 0 && *c < num_cols as isize)
        .map(|(r, c)| grid[r as usize][c as usize])
        .sum()
}

fn process_file(filename: &str) -> io::Result<u32> {
    let mut grid = parse_file(filename)?;
    let num_rows = grid.len();
    let num_cols = grid[0].len();
    grid[0][0] = 1;
    grid[0][num_cols - 1] = 1;
    grid[num_rows - 1][0] = 1;
    grid[num_rows - 1][num_cols - 1] = 1;
    for _ in 0..NUM_STEPS {
        let mut next_grid: Grid = vec![];
        for row in 0..num_rows {
            let mut next_row: Vec<u32> = vec![];
            for col in 0..num_cols {
                let neighbors_on = count_neighbors(&grid, row, col);
                match &grid[row][col] {
                    1 => {
                        if (2..=3).contains(&neighbors_on) {
                            next_row.push(1);
                        } else {
                            next_row.push(0);
                        }
                    }
                    0 => {
                        if neighbors_on == 3 {
                            next_row.push(1);
                        } else {
                            next_row.push(0);
                        }
                    }
                    _ => panic!("unexpected value in grid!"),
                }
            }
            next_grid.push(next_row);
        }
        grid = next_grid;
        grid[0][0] = 1;
        grid[0][num_cols - 1] = 1;
        grid[num_rows - 1][0] = 1;
        grid[num_rows - 1][num_cols - 1] = 1;
    }
    let count = grid.iter().map(|x| x.iter().sum::<u32>()).sum();
    Ok(count)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let progname = &args[0];
    if args.len() < 2 {
        usage(progname);
    }
    let filename = &args[1];
    match process_file(filename) {
        Ok(result) => println!("result = {result}"),
        Err(e) => {
            eprintln!("error while processing file `{filename}': {e}");
            process::exit(1);
        }
    }
}

File: haskell/2015/day18b/src/Main.hs
module Main (main) where

import Control.Monad
import Control.Monad.ST
import Data.Array.ST
import Data.Array.Unboxed
import Data.Int (Int32)
import System.Environment (getArgs, getProgName)
import System.Exit (exitFailure)
import System.IO (hPutStrLn, stderr)
import Control.DeepSeq
import System.Clock

type Grid = UArray (Int, Int) Bool

usage :: String -> IO ()
usage progname = do
    hPutStrLn stderr $ "usage: " ++ progname ++ " <input file>"
    exitFailure

parseInput :: String -> Grid
parseInput content =
    let contentLines = lines content
        numRows = length contentLines
        numCols = length (contentLines !! 0)
        mapChar (row, col) = case (contentLines !! (row - 1)) !! (col -1) of
                                 '#' -> True
                                 '.' -> False
                                 _   -> error ("unexpected char in input at (" ++ show row ++ ", " ++ show col ++ ")")
        initGrid = array ((1, 1), (numRows, numCols)) [((i, j), mapChar (i, j)) | i <- [1..numRows], j <- [1..numCols]]
    in initGrid // [((1, 1), True), ((1, numCols), True), ((numRows, 1), True), ((numRows, numCols), True)]

calcLight :: Grid -> (Int, Int) -> Bool
calcLight grid (row, col) =
    let ((rMin, cMin), (rMax, cMax)) = bounds grid
        onCount = length [() | (dr, dc) <- [(-1, -1), (-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0), (1, 1)],
                               let r = row + dr,
                               let c = col + dc,
                               r >= rMin, r <= rMax,
                               c >= cMin, c <= cMax,
                               grid ! (r, c)]
        lightOn = grid ! (row, col)
    in if lightOn then onCount >= 2 && onCount <= 3
                  else onCount == 3

step :: Grid -> Grid
step grid = runSTUArray $ do
                let gridBounds = bounds grid
                let ((rMin, cMin), (rMax, cMax)) = gridBounds
                mgrid <- newArray gridBounds False :: ST s (STUArray s (Int, Int) Bool)
                forM_ (indices grid) $ \i -> do
                    let newVal = calcLight grid i
                    writeArray mgrid i newVal
                writeArray mgrid (rMin, cMin) True
                writeArray mgrid (rMin, cMax) True
                writeArray mgrid (rMax, cMin) True
                writeArray mgrid (rMax, cMax) True
                return mgrid

runSteps :: Int -> Grid -> Grid
runSteps n grid = iterate step grid !! n

countLights :: Grid -> Int
countLights grid = length $ filter id (elems grid)

process :: String -> Int32
process content =
    let grid = parseInput content
    in fromIntegral $ countLights $ runSteps 100 grid


showTime :: TimeSpec -> String
showTime elapsed =
    let ns = fromIntegral (toNanoSecs elapsed) :: Double
    in if ns < 1000
       then show ns ++ " ns"
       else if ns < 1000000
       then show (ns / 1000.0) ++ " μs"
       else if ns < 1000000000
            then show (ns / 1000000.0) ++ " ms"
            else show (ns / 1000000000.0) ++ " s"
main :: IO ()
main = do
    args <- getArgs
    progname <- getProgName
    case args of
        [filename] -> do
            start <- getTime Monotonic
            content <- readFile filename
            let result = process content
            result `deepseq` return ()
            end <- getTime Monotonic
            let elapsed = diffTimeSpec start end
            putStrLn $ "result = " ++ show result
            putStrLn $ "elapsed time: " ++ showTime elapsed
        _ -> usage progname
